MovieQuiz/Servises/QuestionFactory.swift

```diff
@@ -2,72 +2,108 @@
 //  QuestionFactory.swift
 //  MovieQuiz
 
-
 import Foundation
 
- 
 class QuestionFactory: QuestionFactoryProtocol {
-    
     //MARK: - Variables
+    private let moviesLoader: MoviesLoadingProtocol
+    private var movies: [MostPopularMovie] = []
     weak var delegate: QuestionFactoryDelegate?
     
-    init(delegate: QuestionFactoryDelegate?) {
+    init(delegate: QuestionFactoryDelegate?, moviesLoader: MoviesLoadingProtocol) {
         self.delegate = delegate
+        self.moviesLoader = moviesLoader
     }
     
     //MARK: - Methods
-    func requestNextQuestion() {
-        
-        guard let index = (0..<questions.count).randomElement() else {  // 2
-            delegate?.didRecieveNextQuestion(question: nil)
-            return
+    func loadData() {
+        moviesLoader.loadMovies { result in
```

Comment:
Пропущенное ослабление ссылки здесь - потенциально ошибка в будущем, конкретно в этом варианте не приводит к проблемам, но здесь неявно может захватывать сильный контекст внутри loadMovies. Рекомендую вообще всегда ослаблять ссылку, если не очень понятно надо/не надо -- дешевле же будет :)
```suggestion
        moviesLoader.loadMovies { [weak self] result in
```

Reply:
Этот код из курса, я так понял мы перенесли ссылку на строчку 21, разве нет?
19     func loadData() {
20        moviesLoader.loadMovies { result in
21           DispatchQueue.main.async { [weak self] in
22                guard let self = self else { return }

Reply:
Это нелегкая тема, соглашусь. В учебнике мог остаться и такой вариант, это тоже корректно. 

Я же предлагаю задуматься о том как используются списки захвата, ослабляющие self. 
Еще раз скажу что: это **потенциальное** место ошибки, а не ошибка, в таком коде стоит быть внимательным и несколько раз перепроверить что weak self не нужен. 

Если комплишен функции loadMovies сервис moviesLoader захочет например сильно его сохранить, то внутрь замыкания будет захвачена сильная ссылка на фабрику. Это может приводить к утечкам памяти. Здесь же мы как разработчики этого кода гарантируем, что moviesLoader такого делать не будет, поэтому и ошибки не возникнет. Но если подходить к работке как пользователи сервиса moviesLoader, то лучше думать как "я не знаю как внутри сервис устроен", поэтому захвачу контекст слабо. 

Я предлагаю использовать более безопасный подход к проектированию кода :)

Reply:
Кажется понял, исправил